### PR TITLE
SlackModel for QuadraticModels

### PR DESCRIPTION
**Pull request by geoffroyleconte**
This allows to get H and A faster after using SlackModel. 
I modified qp.data, and created another meta using the code already written in NLPModels.slack_meta.

### CODE REVIEW

**Comment by dpo** [comment]
Great, thank you. Do this work for you in RipQP?

When you've converged on the right design, it would be great to add a test.

**Comment by dpo** [comment]
> Is it a problem if SlackModel changes its argument?

In what sense?

**Comment by geoffroyleconte** [comment]
If the user wants to have the initial QuadraticModel and the QuadraticModel with the slack variables in two different objects.

**Comment by dpo** [comment]
I think the user will often want to keep the original QP separate from the slack QP, so it's fine to copy data as you did.

However, for QPs, it could also be useful to write `SlackModel!`. I like the idea. Could you open an issue and open a separate PR?

**Comment by geoffroyleconte** [comment]
I did not copy qp.data, I only copied qp.meta (I think). What about in this pull request? The two are similar. Or do you mean a pull request in NLPModels?

**Comment by dpo** [comment]
I'm sorry. I didn't realize that you were modifying `qp.data` directly! I think what you implemented here is fine, but it doesn't match `NLPModels.SlackModel` (which doesn't modify the user's problem). You could call your method `SlackModel!` and only make it part of `QuadraticModels`. It would be great to add a few tests though to validate the slack model.

File: src/qpmodel.jl [hunk 1]
```diff
@@ -181,3 +181,38 @@ function NLPModels.jtprod!(qp :: AbstractQuadraticModel, x :: AbstractVector, v
...
   return Atv
 end
+
+function SlackModel!(qp :: AbstractQuadraticModel)
+  qp.meta.ncon == length(qp.meta.jfix) && return qp
+  
+  nfix = length(qp.meta.jfix)
+  ns = qp.meta.ncon - nfix
+  T = eltype(qp.data.c)
+  qp.data.Arows = [qp.data.Arows; qp.meta.jlow; qp.meta.jupp; qp.meta.jrng]
```
Comment (dpo) [inline src/qpmodel.jl:191]: Est-ce qu'il n'y a pas moyen d'utiliser `append!()` ici pour ne pas dupliquer les données ?

**Review by dpo — CHANGES_REQUESTED** [review]

File: src/qpmodel.jl (hunk 1 above)
Comment (geoffroyleconte) [reply src/qpmodel.jl]: Quelque chose comme cela?

File: src/qpmodel.jl [hunk 2]
```diff
@@ -188,10 +188,12 @@ function SlackModel!(qp :: AbstractQuadraticModel)
...
   T = eltype(qp.data.c)
-  qp.data.Arows = [qp.data.Arows; qp.meta.jlow; qp.meta.jupp; qp.meta.jrng]
-  qp.data.Acols = [qp.data.Acols; qp.meta.nvar+1:qp.meta.nvar+ns]
-  qp.data.Avals = [qp.data.Avals; .-ones(T, ns)]
-  qp.data.c = [qp.data.c; zeros(T, ns)]
+  append!(qp.data.Arows, qp.meta.jlow)
+  append!(qp.data.Arows, qp.meta.jupp)
+  append!(qp.data.Arows, qp.meta.jrng)
+  append!(qp.data.Acols, qp.meta.nvar+1:qp.meta.nvar+ns)
+  append!(qp.data.Avals, .-ones(T, ns))
```
Comment (dpo) [inline src/qpmodel.jl:195]: Ici, je me demande si tu ne pourrais pas faire
```julia
append!(qp.data.Avals, (-one(T) for _ = 1 : ns))
```
pour éviter de construire un tableau de -1. Même chose à la ligne suivante.

Comment (geoffroyleconte) [reply src/qpmodel.jl]: Oui, je savais pas qu'on pouvait faire ça, j'ai fait les changements

**Comment by dpo** [comment]
Thank you! 🎉

**Pull request merged**